### PR TITLE
Blackjack EOD

### PR DESCRIPTION
**Pull request by TheArchitectFreeSO**
Introducing Blackjack EOD.

Huge special thanks to Raeven for her expertise on slots and routing.
She also created the dealer object from scratch.

Special thanks to S1ndle for the catalog strings.

Important Note:
While the game (seems to be, hopefully) is 100% working and ready to be
tested in the beta, the dynamic sprites do not yet work. This will be
patched an a future update.

Known Issues:
•Dynamic sprites aren't yet implemented - the table sprites don't
change.
•The dealer won't obey Animate Sim prims to "turn body toward stack
object" so he just faces the table all the time.

<P g="4305B84E" s="15" p="50000" n="Ace Blackjack Table" />

### CODE REVIEW

File: TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs [hunk 1]
```diff
@@ -0,0 +1,2451 @@
...
+            }
+        }
+        /*
+         * Players supplied sound string
+         */
+        private void SoundHandler(string evt, string soundString)
+        {
+            // play the sound
+            HIT.HITVM.Get().PlaySoundEvent(soundString);
+            HIT.HITVM.Get().PlaySoundEvent(soundString);
```
Comment (riperiperi) [inline TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs]: Why is this here twice? The second call will do nothing since two sounds of the same type cannot play at once.

File: TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs [hunk 2]
```diff
@@ -0,0 +1,2451 @@
...
+         */
+        private void InsuranceCallbackHandler(string evt, byte[] insuranceIfNotZero)
+        {
+            int spentForInsurance = BitConverter.ToInt32(insuranceIfNotZero, 0);
+            // if insurance was bought, create a new alert confirming it was bought
+            if (spentForInsurance > 0)
+            {
+                // play the ka-ching sound
+                HIT.HITVM.Get().PlaySoundEvent("ui_object_place");
+                HIT.HITVM.Get().PlaySoundEvent("ui_object_place");
```
Comment (riperiperi) [inline TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs]: Twice again.

File: TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs [hunk 3]
```diff
@@ -0,0 +1,2451 @@
...
+         */
+        private void ChangeMyBetHandler(string evt, byte[] newBetAmount)
+        {
+            // update bet amount
+            MyCurrentBetAmount = BitConverter.ToInt32(newBetAmount, 0);
+            // update displayed bet text below and above
+            UpdateMyBetAmount("" + MyCurrentBetAmount);
+            // play the ka-ching sound
+            HIT.HITVM.Get().PlaySoundEvent("ui_object_place");
+            HIT.HITVM.Get().PlaySoundEvent("ui_object_place");
```
Comment (riperiperi) [inline TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs]: ???

File: TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs [hunk 4]
```diff
@@ -0,0 +1,2451 @@
...
+                {
+                    if (MainPlayerCardContainers.Count == 0)
+                        AddMainCardContainer(0, currentList.ToArray());
+                    else
+                        SetNewActiveHand(0, currentList.ToArray());
+                }
+                // play a sound
+                HIT.HITVM.Get().PlaySoundEvent("blackjack_cards_hit");
+                HIT.HITVM.Get().PlaySoundEvent("blackjack_cards_hit");
+                HIT.HITVM.Get().PlaySoundEvent("blackjack_cards_hit");
```
Comment (riperiperi) [inline TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs]: Three times?

File: TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs [hunk 5]
```diff
@@ -0,0 +1,2451 @@
...
+                appendix = ": " + GameFacade.Strings["UIText", "263", "6"]; // "Hit"
+            else if (evt[11].Equals('t')) // "blackjack_stand_broadcast"
+                appendix = ": " + GameFacade.Strings["UIText", "263", "7"]; // "Stand"
+            else if (evt[11].Equals('p')) // "blackjack_split_broadcast"
+            {
+                appendix = ": " + GameFacade.Strings["UIText", "263", "8"]; // "Split"
+                MakeSplitLabelVisible(player[0]);
+                // play the ka-ching sound
+                HIT.HITVM.Get().PlaySoundEvent("ui_object_place");
+                HIT.HITVM.Get().PlaySoundEvent("ui_object_place");
```
Comment (riperiperi) [inline TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs]: .

File: TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs [hunk 6]
```diff
@@ -0,0 +1,2451 @@
...
+            }
+            else if (evt[11].Equals('o')) // "blackjack_double_broadcast"
+            {
+                appendix = ": " + GameFacade.Strings["UIText", "263", "9"]; // "Double down"
+                // play the ka-ching sound
+                HIT.HITVM.Get().PlaySoundEvent("ui_object_place");
+                HIT.HITVM.Get().PlaySoundEvent("ui_object_place");
```
Comment (riperiperi) [inline TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs]: .

File: TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs [hunk 7]
```diff
@@ -0,0 +1,2451 @@
...
+         * todo: If the LiveMode & EOD invalidation issue is ever solved on the EOD-wide scale, this middle-man method will be unnecessary.
+         */
+        private void SetNewTip(string newTip)
+        {
+            SetTip(newTip);
+            Parent.Invalidate();
+            Parent.Invalidate();
+            Parent.Invalidate();
+            Parent.Invalidate();
+            Parent.Invalidate();
```
Comment (riperiperi) [inline TSOClient/tso.client/UI/Panels/EODs/UIBlackjackEOD.cs]: Calling this 16 times in a row will not do anything differently.

File: TSOClient/tso.simantics/NetPlay/EODs/Handlers/VMEODBlackjackPlugin.cs [hunk 8]
```diff
@@ -0,0 +1,2244 @@
...
+                        break;
+                    }
+                case VMEODBlackjackStates.Finale:
+                    {
+                        int dealerTotal = 0; // send 0 if blackjack
+                        if (!DealerPlayer.CurrentHandType.Equals(VMEODBlackjackHandTypes.Blackjack))
+                            dealerTotal = DealerPlayer.CurrentHandTotal;
+                        Lobby.Broadcast("dealer_hand_total", BitConverter.GetBytes(dealerTotal));
+                        Tock = 0;
+                        IdleTi
```
Comment (riperiperi) [inline TSOClient/tso.simantics/NetPlay/EODs/Handlers/VMEODBlackjackPlugin.cs]: `(Lobby.GetSlotData(player).PendingTransaction)`

Comment (riperiperi) [inline TSOClient/tso.simantics/NetPlay/EODs/Handlers/VMEODBlackjackPlugin.cs]: `(!Lobby.GetSlotData(player).BetSubmitted)`

Comment (riperiperi) [inline TSOClient/tso.simantics/NetPlay/EODs/Handlers/VMEODBlackjackPlugin.cs]: `Lobby.GetSlotData(player).BetAccepted`

Comment (riperiperi) [inline TSOClient/tso.simantics/NetPlay/EODs/Handlers/VMEODBlackjackPlugin.cs]: `Lobby.GetSlotData(player).PromptedForInsurance`

Comment (riperiperi) [inline TSOClient/tso.simantics/NetPlay/EODs/Handlers/VMEODBlackjackPlugin.cs]: IndexOutOfRangeException? I don't like the presence of this function at all, and its use on the background thread. (as discussed)

File: TSOClient/tso.simantics/NetPlay/EODs/Utils/EODLobby.cs [hunk 9]
```diff
@@ -95,10 +95,20 @@ public bool IsFull()
...
+                {
+                    return false;
+                }
+            }
+            return true;
+        }
+
         public bool Join(VMEODClient client, short slot)
         {
-            if (slot == -1) //take first available slot
```
Comment (riperiperi) [inline TSOClient/tso.simantics/NetPlay/EODs/Utils/EODLobby.cs]: This is necessary for trade...

Comment (TheArchitectFreeSO) [reply TSOClient/tso.simantics/NetPlay/EODs/Utils/EODLobby.cs]: For the EODLobby.cs change, I think I just missed that it had been updated for Secure Trade and was working on the older version. Apologies and thanks for patching it.

File: TSOClient/tso.simantics/NetPlay/EODs/Handlers/VMEODBlackjackPlugin.cs (hunk 8 above)
Comment (TheArchitectFreeSO) [reply TSOClient/tso.simantics/NetPlay/EODs/Handlers/VMEODBlackjackPlugin.cs]: Facepalm on the bools
